src/IO/ReadBufferFromPocoSocket.cpp

```diff
@@ -28,10 +28,25 @@ bool ReadBufferFromPocoSocket::nextImpl()
     ssize_t bytes_read = 0;
     Stopwatch watch;
 
+    int flags = 0;
+    if (fiber)
+        flags |= MSG_DONTWAIT;
+
     /// Add more details to exceptions.
     try
     {
-        bytes_read = socket.impl()->receiveBytes(internal_buffer.begin(), internal_buffer.size());
+        bytes_read = socket.impl()->receiveBytes(internal_buffer.begin(), internal_buffer.size(), flags);
+
+        /// If fiber is specified, and read is blocking, run fiber and try again later.
+        /// It is expected that file descriptor may be polled externally.
+        /// Note that receive timeout is not checked here. External code should check it while polling.
+        while (bytes_read < 0 && fiber && errno == EAGAIN)
+        {
+            //fiber->fd = socket.impl()->sockfd();
+            //fiber->timeout = socket.impl()->getReceiveTimeout();
+            *fiber = std::move(*fiber).resume();
```

Comment:
So when the read would block, we (1) resume the caller fiber. Before that, we save into the processor state:
- current call state as a fiber
- socket descriptor for the caller to do epoll
- mark that the processor result is Async

(2) After that, the caller sees that the processor result is Async, and adds the corresponding file descriptor to its epoll set.

(3) When epoll succeeds, we see for which descriptor, and resume the corresponding fiber we created at step (1).

Am I getting this right?

Reply:
Yes.

Also, epoll is executed into PipelineExecutor, in separate thread. When it succeeds, we continue execution of a processor, which continues fiber execution. And next read should return some bytes.